[Screen Time Refactoring] Add WKWebsiteDataStore property for Screen Time history donations

jesxilin opened the pull request:
#### 640cc8483b4c7816045e413c10eca9f2dfe424e2
<pre>
[Screen Time Refactoring] Add WKWebsiteDataStore property for Screen Time history donations
<a href="https://bugs.webkit.org/show_bug.cgi?id=285239">https://bugs.webkit.org/show_bug.cgi?id=285239</a>
<a href="https://rdar.apple.com/140439004">rdar://140439004</a>

Reviewed by Aditya Keerthi and Abrar Rahman Protyasha.

Added WKWebsiteDataStore property for Screen Time history deletions with
identifiers.
Created a new WebsiteDataType::ScreenTime.

Altered WebsiteDataStore::fetchDataAndApply to call Screen Time&apos;s new
API to obtain urls for a certain profile identifier.

Added two new methods:
- removeScreenTimeData corresponds to removeDataOfTypes that takes
in NSArray&lt;WKWebsiteDataRecord *&gt; * dataRecords
- removeScreenTimeDataWithInterval corresponds to removeDataOfTypes that takes
in an NSDate * date

Wrote two new API tests (RemoveDataWithTimeInterval and RemoveData).
RemoveData relies on Screen Time&apos;s API and will need to be changed later on.

* Source/WebCore/PAL/pal/cocoa/ScreenTimeSoftLink.h:
* Source/WebCore/PAL/pal/cocoa/ScreenTimeSoftLink.mm:
* Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.cpp:
(WebKit::WebResourceLoadStatisticsStore::monitoredDataTypes):
* Source/WebKit/Shared/WebsiteData/WebsiteData.cpp:
(WebKit::WebsiteData::ownerProcess):
* Source/WebKit/Shared/WebsiteData/WebsiteDataType.h:
* Source/WebKit/Shared/WebsiteData/WebsiteDataType.serialization.in:
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataRecord.mm:
(dataTypesToString):
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataRecordInternal.h:
(WebKit::toWebsiteDataType):
(WebKit::toWKWebsiteDataTypes):
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataRecordPrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
(+[WKWebsiteDataStore allWebsiteDataTypes]):
* Source/WebKit/UIProcess/WebsiteData/Cocoa/WebsiteDataStoreCocoa.mm:
(WebKit::WebsiteDataStore::removeScreenTimeData):
(WebKit::WebsiteDataStore::removeScreenTimeDataWithInterval):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataRecord.h:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::fetchDataAndApply):
(WebKit::WebsiteDataStore::removeData):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStoreClient.h:
(WebKit::WebsiteDataStoreClient::getScreenTimeURLs const):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm:
(TEST(ScreenTime, RemoveDataWithTimeInterval)):
(TEST(ScreenTime, RemoveData)):

Canonical link: <a href="https://commits.webkit.org/290431@main">https://commits.webkit.org/290431@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6a9ff798a5e0fdab36d18a602d52eb581b6ef625

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/89942 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/9471 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/44841 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/94942 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/40715 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/91994 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/9858 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/17753 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/69247 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/26863 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/92943 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/7551 "Found 1 new test failure: fast/forms/ios/focus-input-in-fixed.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/81600 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/49603 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/7277 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/35977 "Found 6 new test failures: fast/html/process-end-tag-for-inbody-crash.html fullscreen/empty-anonymous-block-continuation-crash.html fullscreen/full-screen-request-removed.html fullscreen/fullscreen-cancel-after-request-crash.html media/modern-media-controls/fullscreen-button/fullscreen-button.html media/modern-media-controls/placard-support/placard-support-airplay-fullscreen.html (failure)") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/39849 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/77620 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/37021 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/96765 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/17130 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/12580 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/78213 "layout-tests (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/17386 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/77419 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/77447 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/21898 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/20490 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/10324 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/14150 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/17140 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/22465 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/16881 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/20333 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/18664 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
```diff
@@ -78,6 +78,11 @@
 #endif
 
+#if ENABLE(SCREEN_TIME)
+#import <ScreenTime/STWebHistory.h>
```
You get this include through `ScreenTimeSoftLink.h` anyway, so it's safe to remove.

pxlcoder commented on Source/WebKit/Shared/WebsiteData/WebsiteDataType.h:
```diff
@@ -50,6 +50,7 @@ enum class WebsiteDataType : uint32_t {
 #endif
     FileSystem = 1 << 19,
     BackgroundFetchStorage = 1 << 20,
+    ScreenTime = 1 << 21,
```
This should probably be guarded by `ENABLE(SCREEN_TIME)`, as should quite a few places in this patch.

pxlcoder commented on Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
```diff
@@ -165,6 +176,29 @@ void didReceiveAuthenticationChallenge(Ref<WebKit::AuthenticationChallengeProxy>
+        if (![PAL::getSTWebHistoryClass() instancesRespondToSelector:@selector(fetchAllHistoryWithCompletionHandler:)])
+            return;
+
+        // Create STWebHistory
```
Remove comment.

aprotyas commented on Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
```diff
@@ -86,6 +91,12 @@ @interface WKWebsiteDataStore (WKWebPushHandling)
 
+#if ENABLE(SCREEN_TIME)
+@interface STWebHistory (Staging_140439004)
+- (void)fetchAllHistoryWithCompletionHandler:(void (^)(NSSet<NSURL *> * urls, NSError *  error))completionHandler;
```
Style guide nit

```suggestion
- (void)fetchAllHistoryWithCompletionHandler:(void (^)(NSSet<NSURL *> *urls, NSError *error))completionHandler;
```

aprotyas commented on Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
```diff
@@ -165,6 +176,29 @@ void didReceiveAuthenticationChallenge(Ref<WebKit::AuthenticationChallengeProxy>
+        STWebHistoryProfileIdentifier const WebHistoryProfileIdentifier = identifer;
+        RetainPtr webHistory = adoptNS([PAL::allocSTWebHistoryInstance() initWithProfileIdentifier:WebHistoryProfileIdentifier]);
+
+        // Use STWebHistory to fetch all website history for identifier
```
These two comments can go.

pxlcoder commented on Source/WebKit/UIProcess/WebsiteData/Cocoa/WebsiteDataStoreCocoa.mm:
```diff
@@ -332,6 +343,35 @@ static String defaultWebsiteDataStoreRootDirectory()
+        return;
+
+    STWebHistoryProfileIdentifier const WebHistoryProfileIdentifierPersonal = configuration().identifier() ? configuration().identifier()->toString() : ""_s;
+    STWebHistory *webHistory = [PAL::allocSTWebHistoryInstance() initWithProfileIdentifier:WebHistoryProfileIdentifierPersonal];
```
Memory leak.

beidson replied:
RetainPtr webHistory = adoptNS(...)

aprotyas commented on Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
```diff
@@ -165,6 +176,29 @@ void didReceiveAuthenticationChallenge(Ref<WebKit::AuthenticationChallengeProxy>
+        RetainPtr webHistory = adoptNS([PAL::allocSTWebHistoryInstance() initWithProfileIdentifier:WebHistoryProfileIdentifier]);
+
+        // Use STWebHistory to fetch all website history for identifier
+        [webHistory fetchAllHistoryWithCompletionHandler:^(NSSet<NSURL *> *urls, NSError *  error) {
```
Style guide nit
```suggestion
        [webHistory fetchAllHistoryWithCompletionHandler:^(NSSet<NSURL *> *urls, NSError *error) {
```

pxlcoder replied:
This should use `makeBlockPtr`, capture the `callback`, and move it.

pxlcoder commented on Source/WebKit/UIProcess/WebsiteData/Cocoa/WebsiteDataStoreCocoa.mm:
```diff
@@ -332,6 +343,35 @@ static String defaultWebsiteDataStoreRootDirectory()
+{
+    STWebHistoryProfileIdentifier const WebHistoryProfileIdentifierPersonal = configuration().identifier() ? configuration().identifier()->toString() : ""_s;
+
+    STWebHistory *webHistory = [PAL::allocSTWebHistoryInstance() initWithProfileIdentifier:WebHistoryProfileIdentifierPersonal];
```
Memory leak.

beidson replied:
`RetainPtr webHistory = adoptNS(...)`

pxlcoder commented on Source/WebKit/UIProcess/WebsiteData/Cocoa/WebsiteDataStoreCocoa.mm:
```diff
@@ -332,6 +343,35 @@ static String defaultWebsiteDataStoreRootDirectory()
+    if (!modifiedSince.isNaN()) {
+        NSTimeInterval timeInterval = modifiedSince.secondsSinceEpoch().seconds();
+        NSDate *date = [NSDate dateWithTimeIntervalSince1970:timeInterval];
+        NSDateInterval *dateInterval = [[NSDateInterval alloc] initWithStartDate:date endDate:NSDate.now];
```
Memory leak.

beidson replied:
RetainPtr = adoptNS(...) also

pxlcoder commented on Source/WebKit/UIProcess/WebsiteData/WebsiteDataRecord.h:
```diff
@@ -32,9 +32,11 @@
 #include <wtf/text/StringHash.h>
 #include <wtf/text/WTFString.h>
 
+
```
Remove.

aprotyas commented on Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
```diff
@@ -165,6 +176,29 @@ void didReceiveAuthenticationChallenge(Ref<WebKit::AuthenticationChallengeProxy>
+
+        // Use STWebHistory to fetch all website history for identifier
+        [webHistory fetchAllHistoryWithCompletionHandler:^(NSSet<NSURL *> *urls, NSError *  error) {
+            UNUSED_PARAM(error);
```
I think we wouldn't surface an unused param warning here anyway, so maybe safe to just remove this? :)

aprotyas commented on Source/WebKit/UIProcess/WebsiteData/Cocoa/WebsiteDataStoreCocoa.mm:
```diff
@@ -57,12 +57,23 @@
 #endif
 
+#if ENABLE(SCREEN_TIME)
+#import <ScreenTime/STWebHistory.h>
```
Same comment about not needing to include this separately.

aprotyas commented on Source/WebKit/UIProcess/WebsiteData/Cocoa/WebsiteDataStoreCocoa.mm:
```diff
@@ -57,12 +57,23 @@
 
+#if ENABLE(SCREEN_TIME)
+@interface STWebHistory (Staging_140439004)
+- (void)fetchAllHistoryWithCompletionHandler:(void (^)(NSSet<NSURL *> * urls, NSError * error))completionHandler;
```
Style
```suggestion
- (void)fetchAllHistoryWithCompletionHandler:(void (^)(NSSet<NSURL *> *urls, NSError *error))completionHandler;
```

pxlcoder commented on Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
```diff
@@ -748,17 +748,24 @@ void WebsiteDataStore::fetchDataAndApply(OptionSet<WebsiteDataType> dataTypes, O
+    if (dataTypes.contains(WebsiteDataType::ScreenTime) && isPersistent()) {
+        auto identifier = configuration().identifier();
+        const String identifierString = identifier ? identifier->toString() : ""_s;
+        m_client->getScreenTimeURLs(identifierString, [&](HashSet<URL> urls) {
```
```suggestion
        m_client->getScreenTimeURLs(identifierString, [callbackAggregator](HashSet<URL> urls) {
```

pxlcoder commented on Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataRecordPrivate.h:
```diff
@@ -39,6 +39,8 @@ WK_EXTERN NSString * const _WKWebsiteDataTypeAdClickAttributions WK_API_AVAILABL
 WK_EXTERN NSString * const _WKWebsiteDataTypeAlternativeServices WK_API_AVAILABLE(macos(11.0), ios(14.0));
 WK_EXTERN NSString * const _WKWebsiteDataTypeFileSystem WK_API_DEPRECATED_WITH_REPLACEMENT("WKWebsiteDataTypeFileSystem", macos(13.0, 14.0), ios(16.0, 17.0));
+WK_EXTERN NSString * const _WKWebsiteDataTypeScreenTime WK_API_AVAILABLE(macos(WK_MAC_TBA), ios(WK_IOS_TBA));
+
```
Remove extra newline.

aprotyas commented on Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
```diff
@@ -283,6 +283,10 @@ class WebsiteDataStore : public API::ObjectImpl<API::Object::Type::WebsiteDataSt
     const WebsiteDataStoreConfiguration::Directories& resolvedDirectories() const;
     FileSystem::Salt mediaKeysStorageSalt() const;
+#if ENABLE(SCREEN_TIME)
+    void removeScreenTimeData(HashSet<URL> websitesToRemove);
```
You probably want to pass an r-value reference or const l-value reference, whichever is more appropriate for callers? Or else you'd be copying the set on every invocation.

aprotyas commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm:
```diff
@@ -434,4 +435,45 @@ - (void)userContentController:(WKUserContentController *)userContentController d
+
+TEST(ScreenTime, RemoveData)
+{
+    // FIXME: Add test once Screen TIme gives fetchAllHistoryWithCompletionHandler API
```
```suggestion
    // FIXME: Add test once Screen Time gives fetchAllHistoryWithCompletionHandler API
```

aprotyas commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm:
```diff
@@ -434,4 +435,45 @@ - (void)userContentController:(WKUserContentController *)userContentController d
+    EXPECT_TRUE(removedHistory);
+}
+
+TEST(ScreenTime, RemoveData)
```
Alternatively you could write the test now and keep it disabled?
```suggestion
TEST(ScreenTime, DISABLE_RemoveData)
```

beidson commented on Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
```diff
@@ -748,17 +748,24 @@ void WebsiteDataStore::fetchDataAndApply(OptionSet<WebsiteDataType> dataTypes, O
             WebsiteData websiteData;
-            websiteData.entries = mediaKeysStorageOrigins(mediaKeysStorageDirectory).map([](auto& origin) {
-                return WebsiteData::Entry { origin, WebsiteDataType::MediaKeys, 0 };
+            // Get origins for all screen time URLs.
```
Don't think this comment is needed.

rr-codes commented on Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
```diff
@@ -165,6 +175,26 @@ void didReceiveAuthenticationChallenge(Ref<WebKit::AuthenticationChallengeProxy>
+    void getScreenTimeURLs(const String& identifer, CompletionHandler<void(HashSet<URL>&&)>&& callback) const final
+    {
+        if (![PAL::getSTWebHistoryClass() instancesRespondToSelector:@selector(fetchAllHistoryWithCompletionHandler:)])
+            return;
```
the completion handler should always be called in all cases (and should probably be named "completionHandler" instead of 'callback"

rr-codes commented on Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
```diff
@@ -165,6 +175,26 @@ void didReceiveAuthenticationChallenge(Ref<WebKit::AuthenticationChallengeProxy>
+        if (![PAL::getSTWebHistoryClass() instancesRespondToSelector:@selector(fetchAllHistoryWithCompletionHandler:)])
+            return;
+
+        STWebHistoryProfileIdentifier const WebHistoryProfileIdentifier = identifer;
```
I don't think `const` is needed. also should be `webHistoryProfileIdentifier` since it's a variable name.

rr-codes commented on Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
```diff
@@ -748,17 +748,23 @@ void WebsiteDataStore::fetchDataAndApply(OptionSet<WebsiteDataType> dataTypes, O
-        protectedQueue()->dispatch([mediaKeysStorageDirectory = crossThreadCopy(WTFMove(mediaKeysStorageDirectory)), callbackAggregator] {
+#if ENABLE(SCREEN_TIME)
+    if (dataTypes.contains(WebsiteDataType::ScreenTime) && isPersistent()) {
+        auto identifier = configuration().identifier();
```
when is identifier null?

jesxilin replied:
Identifier is null unless configuration is `initWithIdentifier`.

pxlcoder commented on Source/WebKit/UIProcess/WebsiteData/Cocoa/WebsiteDataStoreCocoa.mm:
```diff
@@ -332,6 +342,35 @@ static String defaultWebsiteDataStoreRootDirectory()
+    if (![PAL::getSTWebHistoryClass() instancesRespondToSelector:@selector(fetchAllHistoryWithCompletionHandler:)])
+        return;
+
+    STWebHistoryProfileIdentifier webHistoryProfileIdentifier = configuration().identifier() ? configuration().identifier()->toString() : ""_s;
```
```suggestion
    STWebHistoryProfileIdentifier identifier = configuration().identifier() ? configuration().identifier()->toString() : nil;
```

If the identifier is std::nullopt, the `STWebHistoryProfileIdentifier` should be nil, not the empty string. nil is the default, and is what we donate with by default. Using the empty string is a correctness issue.

Also suggest shortening the variable name.

jesxilin replied:
You're right, will change this. Thanks!

pxlcoder commented on Source/WebKit/UIProcess/WebsiteData/Cocoa/WebsiteDataStoreCocoa.mm:
```diff
@@ -332,6 +342,35 @@ static String defaultWebsiteDataStoreRootDirectory()
+
+void WebsiteDataStore::removeScreenTimeDataWithInterval(WallTime modifiedSince)
+{
+    STWebHistoryProfileIdentifier webHistoryProfileIdentifier = configuration().identifier() ? configuration().identifier()->toString() : ""_s;
```
Same comment as above.

pxlcoder commented on Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
```diff
@@ -748,17 +748,23 @@ void WebsiteDataStore::fetchDataAndApply(OptionSet<WebsiteDataType> dataTypes, O
+#if ENABLE(SCREEN_TIME)
+    if (dataTypes.contains(WebsiteDataType::ScreenTime) && isPersistent()) {
+        auto identifier = configuration().identifier();
+        const String identifierString = identifier ? identifier->toString() : ""_s;
```
I think we should pass in `configuration().identifier()` to `getScreenTimeURLs`, and handle this conversion inside there. We also need to make sure to use a `nil` identifier and not an empty one, like the other comments.

jesxilin replied:
That makes sense!

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm:
```diff
@@ -434,4 +435,45 @@ - (void)userContentController:(WKUserContentController *)userContentController d
+
+TEST(ScreenTime, RemoveData)
+{
+    // FIXME: Add test once Screen Time gives fetchAllHistoryWithCompletionHandler API
```
```suggestion
    // FIXME: Add test once Screen Time implements `-[STWebHistory fetchAllHistoryWithCompletionHandler:]`.
```

pxlcoder commented on Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
```diff
@@ -165,6 +175,31 @@ void didReceiveAuthenticationChallenge(Ref<WebKit::AuthenticationChallengeProxy>
         [m_delegate.getAutoreleased() didReceiveAuthenticationChallenge:nsURLChallenge completionHandler:completionHandler.get()];
     }
+#if ENABLE(SCREEN_TIME)
+    void getScreenTimeURLs(const String& identifer, CompletionHandler<void(HashSet<URL>&&)>&& completionHandler) const final
```
```suggestion
    void getScreenTimeURLs(const String& identifier, CompletionHandler<void(HashSet<URL>&&)>&& completionHandler) const final
```

jesxilin replied:
I searched for `identifer` and so many results came up... oh my

pxlcoder commented on Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
```diff
@@ -165,6 +175,31 @@ void didReceiveAuthenticationChallenge(Ref<WebKit::AuthenticationChallengeProxy>
 
         [m_delegate.getAutoreleased() didReceiveAuthenticationChallenge:nsURLChallenge completionHandler:completionHandler.get()];
     }
+#if ENABLE(SCREEN_TIME)
```
```suggestion

#if ENABLE(SCREEN_TIME)
```

pxlcoder commented on Source/WebKit/UIProcess/WebsiteData/Cocoa/WebsiteDataStoreCocoa.mm:
```diff
@@ -332,6 +342,35 @@ static String defaultWebsiteDataStoreRootDirectory()
     });
 }
 
+#if ENABLE(SCREEN_TIME)
```
```suggestion
#if ENABLE(SCREEN_TIME)

```

pxlcoder commented on Source/WebKit/Shared/WebsiteData/WebsiteDataType.h:
```diff
@@ -78,6 +81,9 @@ template<> struct EnumTraitsForPersistence<WebKit::WebsiteDataType> {
+#endif
+#if ENABLE(SCREEN_TIME)
+        WebKit::WebsiteDataType::ScreenTime,
 #endif
```
I think this needs be kept in order of the enum.

pxlcoder commented on Source/WebKit/Shared/WebsiteData/WebsiteDataType.serialization.in:
```diff
@@ -42,7 +42,10 @@ headers: "WebsiteDataType.h"
+#endif
+#if ENABLE(SCREEN_TIME)
+    ScreenTime,
 #endif
```
I think this needs to be kept in order of the enum.

aprotyas commented on Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
```diff
@@ -166,6 +176,35 @@ void didReceiveAuthenticationChallenge(Ref<WebKit::AuthenticationChallengeProxy>
+
+        RetainPtr webHistory = adoptNS([PAL::allocSTWebHistoryInstance() initWithProfileIdentifier:webHistoryProfileId]);
+
+        [webHistory fetchAllHistoryWithCompletionHandler:makeBlockPtr([completionHandler = WTFMove(completionHandler)] (NSSet<NSURL *> *urls, NSError *error) mutable {
```
Style guide; no space between capture list and arg list.

https://webkit.org/code-style-guidelines/#spacing-lambda-paren

```suggestion
        [webHistory fetchAllHistoryWithCompletionHandler:makeBlockPtr([completionHandler = WTFMove(completionHandler)](NSSet<NSURL *> *urls, NSError *error) mutable {
```

pxlcoder commented on Source/WebKit/UIProcess/WebsiteData/WebsiteDataStoreClient.h:
```diff
@@ -119,6 +119,10 @@ class WebsiteDataStoreClient {
     {
     }
 
+    virtual void getScreenTimeURLs(std::optional<WTF::UUID>, CompletionHandler<void(HashSet<URL>&&)>&& completionHandler) const
```
```suggestion
    virtual void getScreenTimeURLs(std::optional<WTF::UUID>, CompletionHandler<void(HashSet<URL>&&)>&&) const
```

pxlcoder commented on Source/WebKit/UIProcess/WebsiteData/Cocoa/WebsiteDataStoreCocoa.mm:
```diff
@@ -332,6 +342,41 @@ static String defaultWebsiteDataStoreRootDirectory()
+    if (![PAL::getSTWebHistoryClass() instancesRespondToSelector:@selector(fetchAllHistoryWithCompletionHandler:)])
+        return;
+
+    STWebHistoryProfileIdentifier webHistoryProfileId = nil;
```
```suggestion
    STWebHistoryProfileIdentifier profileIdentifier = nil;
```

pxlcoder replied:
(WebKit style guide says to use full words)

pxlcoder commented on Source/WebKit/UIProcess/WebsiteData/Cocoa/WebsiteDataStoreCocoa.mm:
```diff
@@ -332,6 +342,41 @@ static String defaultWebsiteDataStoreRootDirectory()
+
+void WebsiteDataStore::removeScreenTimeDataWithInterval(WallTime modifiedSince)
+{
+    STWebHistoryProfileIdentifier webHistoryProfileId = nil;
```
```suggestion
    STWebHistoryProfileIdentifier profileIdentifier = nil;
```

pxlcoder commented on Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
```diff
@@ -166,6 +176,35 @@ void didReceiveAuthenticationChallenge(Ref<WebKit::AuthenticationChallengeProxy>
+        if (![PAL::getSTWebHistoryClass() instancesRespondToSelector:@selector(fetchAllHistoryWithCompletionHandler:)])
+            return completionHandler({ });
+
+        STWebHistoryProfileIdentifier webHistoryProfileId = nil;
```
```suggestion
        STWebHistoryProfileIdentifier profileIdentifier = nil;
```

jesxilin committed:
[Screen Time Refactoring] Add WKWebsiteDataStore property for Screen Time history donations
https://bugs.webkit.org/show_bug.cgi?id=285239
rdar://140439004

Reviewed by Aditya Keerthi and Abrar Rahman Protyasha.

Added WKWebsiteDataStore property for Screen Time history deletions with
identifiers.
Created a new WebsiteDataType::ScreenTime.

Altered WebsiteDataStore::fetchDataAndApply to call Screen Time's new
API to obtain urls for a certain profile identifier.

Added two new methods:
- removeScreenTimeData corresponds to removeDataOfTypes that takes
in NSArray<WKWebsiteDataRecord *> * dataRecords
- removeScreenTimeDataWithInterval corresponds to removeDataOfTypes that takes
in an NSDate * date

Wrote two new API tests (RemoveDataWithTimeInterval and RemoveData).
RemoveData relies on Screen Time's API and will need to be changed later on.

* Source/WebCore/PAL/pal/cocoa/ScreenTimeSoftLink.h:
* Source/WebCore/PAL/pal/cocoa/ScreenTimeSoftLink.mm:
* Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.cpp:
(WebKit::WebResourceLoadStatisticsStore::monitoredDataTypes):
* Source/WebKit/Shared/WebsiteData/WebsiteData.cpp:
(WebKit::WebsiteData::ownerProcess):
* Source/WebKit/Shared/WebsiteData/WebsiteDataType.h:
* Source/WebKit/Shared/WebsiteData/WebsiteDataType.serialization.in:
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataRecord.mm:
(dataTypesToString):
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataRecordInternal.h:
(WebKit::toWebsiteDataType):
(WebKit::toWKWebsiteDataTypes):
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataRecordPrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
(+[WKWebsiteDataStore allWebsiteDataTypes]):
* Source/WebKit/UIProcess/WebsiteData/Cocoa/WebsiteDataStoreCocoa.mm:
(WebKit::WebsiteDataStore::removeScreenTimeData):
(WebKit::WebsiteDataStore::removeScreenTimeDataWithInterval):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataRecord.h:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::fetchDataAndApply):
(WebKit::WebsiteDataStore::removeData):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStoreClient.h:
(WebKit::WebsiteDataStoreClient::getScreenTimeURLs const):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm:
(TEST(ScreenTime, RemoveDataWithTimeInterval)):
(TEST(ScreenTime, RemoveData)):

Canonical link: https://commits.webkit.org/290431@main
